### ISSUE TITLE
Incorporate Posit Support

### ISSUE THREAD

**Issue by Afonso-2403** [pytorch/pytorch#57724]
## 🚀 Feature
Incorporate support for Posit number format as another dtype of PyTorch

## Motivation
The Posit number format (introduced in [Beating Floating Point at its Own Game](http://www.johngustafson.net/pdfs/BeatingFloatingPoint.pdf)) has been sparking a lot of interest in the research community lately, in particular for its potential on low precision training and inference of Neural Networks. 
Researchers have been building their own frameworks for deep learning on top of Posit, some based on PyTorch (such as [PositNN](https://github.com/hpc-ulisboa/posit-neuralnet)), others on TensorFlow (such as [Deep PeNSieve](https://github.com/RaulMurillo/deep-pensieve)).
However, it would be much nicer to have PyTorch support Posit, so that it could be used directly for research without the need to build a framework from scratch.

## Ideas For Implementation
Following [Issue#52673](https://github.com/pytorch/pytorch/issues/52673) and after some discussion with @RaulMurillo, I believe that the best way to do this is by having different Posit types be supported as another dtype, backed by an external library such as [Universal](https://github.com/stillwater-sc/universal) and register kernel implementations to the dispatcher for each method.
[Issue#755](https://github.com/pytorch/pytorch/issues/755) and [Issue#33152](https://github.com/pytorch/pytorch/issues/33152) hold a history of the process that complex numbers went through until they became supported, and I was thinking that a similar process could be undergone for Posits.
Since I am quite inexperienced in PyTorch and its internals, I would love to have some insights on this suggestion: if it makes sense to go this way, if there are obstacles that I should be aware of, etc.
@albanD @ezyang (and other people interested), is this a good way forward?

**Comment by ezyang** [comment]
cc @wickedfoo

**Comment by ezyang** [comment]
We did successfully let people prototype complex number support out of tree but a lot of it didn't make sense because you couldn't reuse existing TensorIterator scaffolding. I think if someone wanted to embark on this I'd suggest doing so on a fork of PyTorch instead, it will just be easier. Add a new dtype and work on bringup.

I'm not too certain how serious we can get about posits in core. I think an important question for us is how soon hardware acceleration for it is in the future.

**Comment by RaulMurillo** [comment]
@ezyang the problem when working on a fork of PyTorch is the lack or difficulty of maintenance. The framework that @Afonso-2403 mentioned, Deep PeNSieve, is based on an outdated fork of TensorFlow whose internals were modified ad-hod to support posits. However, attempting to incorporate subsequent commits results in numerous merge incompatibilities. I'm afraid something like this could happen if working on a fork of PyTorch.
On the other hand, you mention that doing so will be an easier way to incorporate such new types. Why is this? Which should be the procedure to follow? Of course, any other suggestions for adding support for new types to PyTorch are welcome.

About incorporating posits in the core, I guess that, at this time, it would be preferable to develop an independent module for this type, relying on one of the aforementioned libraries. Even the research community is getting promising results for this type (16-bit posits behave as good as 32-bit floats for training, and 8-bit posits could be used for inference rather than bfloat16 or half-precision), most of the presented works emulate posit computations via software-emulation due to the current lack of hardware support. It is likely that some hardware accelerators for posit arithmetic will appear on the scene in the near future, but currently, this is more a research topic.

All in all, I guess there is an increasing interest in posit arithmetic as an alternative for the floating-point format, especially for its use in Deep Learning, and incorporating this to PyTorch would be a great feature.

**Comment by ezyang** [comment]
> On the other hand, you mention that doing so will be an easier way to incorporate such new types. Why is this?

For TensorIterator operators, it will be the difference between writing one line of code to add posit support, versus having to copy paste 100s of lines (TensorIterator ends up instantiating templated kernels for N scalar types. If you can hook into this instantiation you can put in Posit and let operator overloading do the job. If you are out of tree, there's no external extension hook for this and you have to copy paste all of the logic).

I agree though that keeping a fork up to date is not easy. Perhaps there can be some compromise, e.g., maybe we land a minimal amount of code in highly trafficked areas so that merge conflicts don't develop there, and then let the rest develop in a fork.

**Comment by RaulMurillo** [comment]
> Perhaps there can be some compromise, e.g., maybe we land a minimal amount of code in highly trafficked areas so that merge conflicts don't develop there, and then let the rest develop in a fork.

That would be great!
> For TensorIterator operators, it will be the difference between writing one line of code to add posit support, versus having to copy paste 100s of lines

From what I've read, this should be similar to using Array Iterator API at NumPy. However, I have no experience with PyTorch internals. After a quick search, I guess the majority of modifications should be done in [aten/src/ATen](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen) and [c10/core/ScalarType.h](https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h), but I'm not sure about the procedure to follow. Is there any implementation we can follow as an example? Perhaps by imitation of Half/BFloat16 scalar types, since these, as well as posits, are not included by default in C++?

**Comment by ezyang** [comment]
Check out #21522 #21523 #21860 for the last time we added a new dtype (bfloat16)

**Comment by ezyang** [comment]
#44678 seems to be the most recent time we added a new scalar type. CC @supriyar

**Comment by Afonso-2403** [comment]
Thanks for the suggestions @ezyang , [#44678](https://github.com/pytorch/pytorch/pull/44678) helps in understanding the logic of adding a new dtype.

> Perhaps there can be some compromise, e.g., maybe we land a minimal amount of code in highly trafficked areas so that merge conflicts don't develop there, and then let the rest develop in a fork.

This would be great!
Contrary to quantized tensors, that are built from preexisting floating point ones, Posit Tensors need to have exposed in Python the Tensor creating operations, so what does minimal amount of support mean in this initial commit?

I am still a bit overwhelmed by the levels of indirection of Tensor implementation and Dispatchers, but for these creation operations to be supported also for this new dtypes they should be included in `AT_DISPATCH_ALL_TYPES`, instead of a specific dispatcher like the one for quantized tensors, right?
[#21860](https://github.com/pytorch/pytorch/pull/21860/files?file-filters%5B%5D=.h) seems to have done this by replacing `AT_DISPATCH_ALL_TYPES_AND2` with `AT_DISPATCH_ALL_TYPES_AND3` in the relevant operations.

[This talk](http://blog.ezyang.com/2019/05/pytorch-internals/) really helps understanding the internal architecture, [#21523](https://github.com/pytorch/pytorch/pull/21523/files?file-filters%5B%5D=) and [#21860](https://github.com/pytorch/pytorch/pull/21860/files?file-filters%5B%5D=.h) declare bfloat16 but the code seems mostly legacy as you mentioned, so not much seems to be useful for this task.
As @RaulMurillo mentioned, [c10/core/ScalarType.h](https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h) is where the new dtypes should be declared, [aten/src/ATen](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen) is where the kernels for the operations should be defined, [torch/csrc/utils/](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/) is what exposes to Python, but I am still confused with the Tensor Implementation, relation with Caffe and TensorIterator interfaces, do you have more pointers that can help get a better understanding on where/how to hook up on the internals?

Thanks!

**Comment by ezyang** [comment]
> Posit Tensors need to have exposed in Python the Tensor creating operations, so what does minimal amount of support mean in this initial commit?

I don't think there is too much that is heavily trafficked. Just these two:

1. Add it to ScalarType (because everything else needs this)
2. Legacy storage Python bindings (as this is most likely to get nuked in the near future https://github.com/pytorch/pytorch/issues/47442 ); this is done from codegen and some manually written stuff in torch/csrc

An alternative is to just wait until the legacy storage Python bindings go away, and then I suspect everything can just be in the fork; the legacy storage bindings (along with TH/THC codegen) are the only pieces that are biggest surgery, but because TH/THC codegen is inextricably tied to the implementations of posit it probably has to be done in the fork anyway.

One thing that I'm trying hard not to land in mainline is user facing Python API for posits; ideally this lives in the fork so that people don't go stumbling onto a feature that doesn't work.

> I am still a bit overwhelmed by the levels of indirection of Tensor implementation and Dispatchers, but for these creation operations to be supported also for this new dtypes they should be included in AT_DISPATCH_ALL_TYPES, instead of a specific dispatcher like the one for quantized tensors, right?

Since you are going to have to work to incrementally add support for posits as you build out operation support, you'll probably start off by editing AT_DISPATCH_ALL_TYPES call sites one by one to add posit to the AND list.

> I am still confused with the Tensor Implementation, relation with Caffe and TensorIterator interfaces, do you have more pointers that can help get a better understanding on where/how to hook up on the internals?

I think a good start is to `git grep -i bfloat16` which will give you a decent sense for what parts of the codebase you're going to end up editing. This will roughly break down into (1) tons of operator specific support, (2) utility stuff that you can edit to adjust for posits. No changes to caffe2 should be needed. A quick smattering:

* aten/src/ATen/AccumulateType.h - if you do a sum on posits, what dtype should the posits accumulate into? (Typical floating point you want a higher precision, I remember @wickedfoo telling me about how you do something special with posits but I don't remember what it was)
* aten/src/ATen/DLConvertor.cpp - dlpack support for posits
* aten/src/ATen/NumericUtils.h - stuff like isnan support on numeric types
* c10/core/ScalarType.h - the enum with all the dtypes
* c10/util/BFloat16.h - the class representing a bfloat16 number
* c10/util/MathConstants.h - define pi in your dtype
* c10/util/copysign.h - define copysign on your dtype
* torch/csrc/Storage.h - typed storage for bfloat (there's also a cuda one too)
* torch/csrc/utils/byte_order.cpp - serializing bfloat16 for wire
* torch/csrc/utils/tensor_dtypes.cpp - python dtype bindings

**Comment by Afonso-2403** [comment]
Thanks for the smattering @ezyang, bfloat support seems to be analagous to what should be done for Posit.

> 2\. Legacy storage Python bindings (as this is most likely to get nuked in the near future [Make Python Storage untyped #47442](https://github.com/pytorch/pytorch/issues/47442) ); this is done from codegen and some manually written stuff in torch/csrc

From what I understand, [#59671](https://github.com/pytorch/pytorch/pull/59671) will remove some of the legacy storage, after this, does TH/THC code still need to be addressed?

> An alternative is to just wait until the legacy storage Python bindings go away, and then I suspect everything can just be in the fork;

After some discussion with @RaulMurillo, this seems to be the best approach for now, given that implementations might vary substantially with the external library that is chosen to emulate Posits with software.

The header only library [universal](https://github.com/stillwater-sc/universal), used to emulate Posits, has code that is only available after C++17, namely the usage of [constexpr if statements](https://en.cppreference.com/w/cpp/language/if). However, in PyTorch's [CMakeLists.txt](https://github.com/pytorch/pytorch/blob/54673fc944110a615bf4eadb231bd4bb114ea22f/CMakeLists.txt#L38), it is explicitly said that it requires -std=c++14.
Can I build PyTorch with -std=c++17 or will that break something? If so, is it just with `set(CMAKE_CXX_STANDARD 17)` in the CMakeLists.txt or other patches need to be applied?
Sorry if this is a silly question, but I haven't found a way to circumvent this problem and am a bit stuck on this.

Thanks!

**Comment by ezyang** [comment]
> From what I understand, #59671 will remove some of the legacy storage, after this, does TH/THC code still need to be addressed?

Yes

> Can I build PyTorch with -std=c++17 or will that break something? If so, is it just with set(CMAKE_CXX_STANDARD 17) in the CMakeLists.txt or other patches need to be applied?

This should work but you'll need to find all the spots where 14 is hard coded. We switched to c++14 in #30406

**Comment by Afonso-2403** [comment]
> Technically yes, but there is very very little code at this point; TH is completely dead on master, so it's only THC (and if you're not going to do CUDA support at the beginning, you can skip it)

Great!


> This should work but you'll need to find all the spots where 14 is hard coded. We switched to c++14 in #30406

Thanks, I ended up altering the third party library to make it compliant with c++14.

**Comment by Afonso-2403** [comment]
With the goal of supporting the training and inference of a model, I have been cherry-picking the functions needed to build a Le-Net5 model to train on MNIST data.

When supporting CrossEntropyLoss for the new posit dtype, I am facing the problem of the loss being rounded to the closest integer (0 for common loss values), despite the underlying saved data being the correct value.
On a gdb session, at the end of `at::native::(anonymous namespace)::nll_loss_out_frame`:
```
scalar_t* output_data = output.data_ptr<scalar_t>();
*output_data = output_val;

(gdb) p (*output_data).to_float()
$1 = 0.0377807617
(gdb) torch-tensor-repr output
Python-level repr of output:
tensor(0, dtype=torch.posit16es2)
```

And the output value that is propagated for further functions is 0.

It seems to be specific of Tensors with a single element, since the same read/write code is being used elsewhere and values are correctly represented.
Any clues on why this might be the case?
How can I further debug to understand where is the problem?

Thanks!

**Comment by ezyang** [comment]
Are you sure this isn't just a printing problem? It sure looks like a printing problem to me...

**Comment by Afonso-2403** [comment]
It doesn't look like so, because what is propagated to the next stages is 0.
With example data:

```
>>> criterion = nn.CrossEntropyLoss()
>>> criterion(y_hat_float, y_true)
tensor(0.0378)
>>> y_hat_posit = y_hat_float.to(torch.posit16es2)
>>> criterion(y_hat_posit, y_true)
tensor(0, dtype=torch.posit16es2)
```

I am having some problems with the training part, so I can't yet verify if in the optimizer step the weights are being correctly updated, but since what is exposed in Python is always 0 there seems to be a problem here.

Since this is happening on a single element tensor, could this be a problem with the Scalar class for Posits?
Going step by step in gdb is a bit nightmarish, are there any other debug strategies?

Just for it to be exposed here, work is being conducted in [this fork](https://github.com/Afonso-2403/pytorch) under branch [add-first-posit-dtype](https://github.com/Afonso-2403/pytorch/tree/add-first-posit-dtype).

**Comment by ezyang** [comment]
I guess I'd ask you to humor me (that it's a printing problem), dump out the binary representation of the scalar, and check that it is in fact zero.

One reason why I am pretty sure about this being a printing problem is your gdb session above, where you directly checked the pointer and the data is correct. Scalar tensors are represented the same way as tensors: you get a full TensorImpl and everything. We also have a concept of "Scalar" (things there are stored in double precision) but I don't think this is relevant for your purposes.

One final thought; in PyTorch, we generally assume that float can adequately express the entire space of half precision floating point numbers adequately, and ditto double with float. This is probably not actually true for posits which are constructed totally differently, so you may need to convince PyTorch to not do conversions here. However, 0.0378 is "big" enough that I don't expect it to truncate to zero.

**Comment by Afonso-2403** [comment]
Indeed it looks like a printing problem, the correct values are stored:

```
>>> tensor_1 = torch.tensor(0.0378, dtype=torch.posit16es2)
>>> tensor_1
tensor(0, dtype=torch.posit16es2)
>>> float(tensor_1)
0.037811279296875
>>> tensor_2 = torch.tensor(0.963, dtype=torch.posit16es2)
>>> tensor_2
tensor(0, dtype=torch.posit16es2)
>>> tensor_1 + tensor_2
tensor(1, dtype=torch.posit16es2)
```

However, with `item()` what is retrieved is the int value:
```
>>> tensor_1.item()
0
>>> float(tensor_1.item())
0.0
```

There seems to be some problem with how the Scalar class is treating Posits, because with bigger Tensors everything works smoothly:
```
>>> tensor_3 = torch.tensor([0.0378, 0.987], dtype=torch.posit16es2)
>>> tensor_3
tensor([0.037811279296875, 0.987060546875], dtype=torch.posit16es2)
```

> This is probably not actually true for posits which are constructed totally differently, so you may need to convince PyTorch to not do conversions here.

I'm looking into what might be wrong with the class, can you elaborate further on what are these conversions?

**Comment by ezyang** [comment]
OK, you know what, I agree, it's probably a problem with Scalar.

You have:

```
#define DEFINE_IMPLICIT_CTOR(type, name) \
  Scalar(type vv) : Scalar(vv, true) {}

  AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, DEFINE_IMPLICIT_CTOR)
  AT_FORALL_COMPLEX_TYPES(DEFINE_IMPLICIT_CTOR)
  AT_FORALL_POSIT_TYPES(DEFINE_IMPLICIT_CTOR)

#undef DEFINE_IMPLICIT_CTOR
```

this will turn into

```
  Scalar(posit16es2 vv) : Scalar(vv, true) {}
```

the inner constructor will hit

```
  template <
      typename T,
      typename std::enable_if<
          !std::is_integral<T>::value && !c10::is_complex<T>::value,
          bool>::type* = nullptr>
  Scalar(T vv, bool) : tag(Tag::HAS_d) {
    v.d = convert<decltype(v.d), T>(vv);
  }
```

so the posit will be stored in double precision, and then conversion is done by `convert` template function. I'd check for errors anywhere along those lines.
